docs/proposals/component-registration.md

```diff
@@ -0,0 +1,417 @@
+  - Update how Endpoints are managed
+    - EITHER update the EndpointsController to manage Endpoints based on both Pod and ExternalServiceTarget status
+    - OR modify Endpoints to be a read-only view of service providers (Pod/ExternalServiceTarget) and delete the Endpoints controller
+2. Register component ExternalServiceTargets
+  - Update all components to use consistent namespacing and component labeling
+  - EITHER update "external" components to self-register their ExternalServiceTarget
+  - OR update deployment automation to register deployed components as ExternalServiceTargets
+3. Rewrite ComponentStatuses
+  - Dynamically display a list of components and their readiness (filtered Pod/ExternalServiceTarget or Endpoints query) on `/componentstatuses`
+4. Add Service Status Conditions
+  - Add Service condition policies (all, half+1, at least n) that can be specified for each Service
+  - Add (or modify) a controller to update the Service status conditions based on aggregate provider (Pod/ExternalServiceTarget) status
+  - Make sure Services can be filtered by status conditions (e.g. all ready services with label x)
+  - Deprecate ComponentStatuses and move users to filtered "component" querying of Service Status or Pod/ExternalServiceTarget (if they need the status of individual components, not just the aggregates)
+5. Expand Service or add new a new API resource to allow discovery of API group providers
+
+
+### API Changes
+
+1. Add new namespaced API resource: ExternalServiceTarget
+  - See [Alternate Naming](#alternate-naming)
+  - Similar to Pods
+    - Listable, like most other top-level API resources
+    - Have both Spec and Status (filterable on both/either)
+    - Spec would have (optional) LivenessProbe and ReadinessProbe for determining the condition
+    - Describes a single instance
+    - Can be selected by labels, metadata, spec, and status
+    - Is namespaced (components would use a specific namespace, likely "kube-system")
+    - Status would have an array of [Conditions](../devel/api-conventions.md#typical-status-properties)
+    - Can be named with a name generator (similar to pods created by replication controllers) when instances of the same type don't already have unique IDs (generated name is returned in the create response)
+  - Similar to Endpoints (current External Endpoints usage)
+    - Spec would have address (ip or host) and ports (optionally named) to connect with an external service provider instance
+    - Manually added for non-component External Service provider instances (e.g. Oracle DB) that do not know about k8s (rather than adding Endpoints)
+  - Different from Endpoints
+    - Singular, instead of plural (use -List, namespacing, and labels for grouping)
+    - Includes both Ready & NotReady endpoints
+    - Supports multiple and future conditions (in an extensible Conditions array)
+2. Add a new field ServiceSpec.TargetSelector that selects a subset of ExternalServiceTargets
```

Comment:
Something about the name `TargetSelector` is not sitting right with me. I think it's maybe that if all you do is add `TargetSelector`, you're still left with the original `Selector` field which selects against pods. If we're going to have 2 fields, I'd like the names to be clearer as to what their purpose is. For example, rename `Selector` to `PodSelector` (with backwards compat/conversions) and then have another field for selecting against `ExternalServiceTargets` (I think `TargetSelector` is missing the "external" aspect - maybe something like `ExternalTargetSelector`?).

Although I think I'm leaning more towards just having a single `Selector` that selects both pods and external service targets.